### ISSUE TITLE
Kubernetes DNS resolution fial

### ISSUE THREAD

**Issue by JindrichPilar** [pgbouncer/pgbouncer#332]
I am having problems trying to use pgbouncer in (Azure) Kubernetes, it cannot resolve DNS.

It works on my local machine using docker-compose without problem.

I used custom alpine based image with [alpine pgbouncer package](https://pkgs.alpinelinux.org/package/edge/community/x86/pgbouncer), I also tried docker compiled package by [Edoburu](https://github.com/edoburu/docker-pgbouncer). Both of them use c-ares for domain name resolution (as mentioned in https://github.com/pgbouncer/pgbouncer/issues/122).

The error message I get is:
```
LOG S-0x5611d3b0df90: my-db/root@(bad-af):0 closing because: server dns lookup failed (age=0)
```

I checked domain resolution using ``host`` command from [bind-tools](https://pkgs.alpinelinux.org/package/edge/main/x86/bind-tools) and the same domain name inside same Kuberentes pod is resolved without an issue.

**Comment by cpressland** [comment]
@JindrichPilar - we ran into a similar thing on Azure which was resolved by forcing the pod to use TCP for DNS instead of UDP. You're probably having similar issues in other pods but maybe haven't noticed yet.

Try adding this to the pod spec:

```
      dnsConfig:
        options:
        - name: use-vc
        - name: single-request-reopen
      dnsPolicy: ClusterFirst
```

We're still seeing a huge volume of issues with Postgres on Azure, but not specific to what you're seeing above.

**Comment by JindrichPilar** [comment]
It started working without any changes on my part, no idea why.

@cpressland  Thank You for your response, today I got to try your changes, but Azure CLI forbids me using dnsConfig with ``` Forbidden: DNSConfig: custom pod DNS is disabled by feature gate```

**Comment by petere** [comment]
So is it working now or not?

**Comment by JindrichPilar** [comment]
It is working. Sorry for not closing the issue.

**Comment by PierreF** [comment]
I've also hit this issue and spend some time finding the cause. In my finding this issue should not happen in production and was due to bad DNS server setup.

tl;dr; I was my minikube environment that had "bad" DNS behavior. It returned ServerFail instead of NXDomain from some DNS query. In my case a fix was to configure a real DNS in the /etc/resolv.conf on the physical server

Long version:

* pgbouncer is configured to connect to "pg-host" which is an IPv4 host.
* The actual DNS name is "pg-host.default.svc.cluster.local." and should be resolved because /etc/resolv.conf contains search keyword.
* pgbouncer (actually c-ares) start by doing IPv6 query (AAAA).
  * It start with the full DNS name (pg-host.default.svc.cluster.local.) which fail with NXDomain since it's an IPv4
  * It continue each search domain from resolv.conf until pg-host.cluster.local. which still fail with NXDomain
* The very last DNS name queried in IPv4 is "pg-host.". Which is not handled by Kubernetes DNS itself and it thus forwarded to upstream DNS server
  * In my base (miniukube using kvm2), the dnsmasq of libvirtd
  * Which use the /etc/resolv.conf from the physical server, which reference the 127.0.0.53 (systemd-resolved)
  * and... systemd-resolved will return ServerFail for unqualified DNS query :(
* Since c-ares receive a ServerFail (and not an NXDomain) for it's query, it won't continue and fail

Settings a read DNS server (like 8.8.8.8) in the /etc/resolv.conf from physical host by-pass systemd-resolved.
